File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (!root) return true;

        bool leftValid = isValidBST(root->left);
        if (root->val <= last_val) return false;
        last_val = root->val;
        bool rightValid = isValidBST(root->right);

        return leftValid && rightValid;
    }

private:
    int last_val = std::numeric_limits<int>::min();
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_diameter = 0

        def DFS(root):
            nonlocal max_diameter

            if not root: return -1

            left = 1 + DFS(root.left)
            right = 1 + DFS(root.right)

            max_diameter = max(max_diameter, left + right)

            return max(left, right)

        DFS(root)
        return max_diameter